src/main/java/com/google/edith/servlets/UserInsights.java

```diff
@@ -0,0 +1,217 @@
+package com.google.edith.servlets;
+
+import com.google.appengine.api.datastore.DatastoreService;
+import com.google.appengine.api.datastore.DatastoreServiceFactory;
+import com.google.appengine.api.datastore.Entity;
+import com.google.appengine.api.datastore.EntityNotFoundException;
+import com.google.appengine.api.datastore.FetchOptions;
+import com.google.appengine.api.datastore.Key;
+import com.google.appengine.api.datastore.PreparedQuery;
+import com.google.appengine.api.datastore.Query;
+import com.google.appengine.api.datastore.Query.Filter;
+import com.google.appengine.api.datastore.Query.FilterOperator;
+import com.google.appengine.api.datastore.Query.FilterPredicate;
+import com.google.common.collect.ImmutableMap;
+import com.google.gson.Gson;
+import com.google.gson.JsonObject;
+import java.time.format.DateTimeFormatter;
+import java.time.format.FormatStyle;
+import java.time.LocalDate;
+import java.time.temporal.ChronoUnit;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+/**
+ * This class provides the functionality to parse specifc user data from 
+ * datastore and send an agreggate of that information in a JSON file.
+ */
+public final class UserInsights {
+  
+  private String userId;  
+  private DatastoreService datastore;
+  private static final DateTimeFormatter DATE_FORMATTER = DateTimeFormatter.ofPattern("yyyy-MM-dd");
+  private static final Gson GSON = new Gson();
+
+  public UserInsights(String userId) {
+    this.userId = userId;
+    this.datastore = DatastoreServiceFactory.getDatastoreService();
+  }
+
+  /** This should only be called each time a new user makes an accout. */
+  public void createUserStats() { 
+     List<Key> items = new ArrayList<>();
+
+     Entity userStats = new Entity("UserStats");
+     userStats.setProperty("userId", userId);
+     userStats.setProperty("Items", items);
+     datastore.put(userStats);
+  }
+
+  /** 
+   * Updates the Items list in the UserStats Entity in datastore
+   * corresponding to this user.
+   * @param items - Non-null list of item Keys to be added to the current Item list.
+   */
+  public void updateUserStats(List<Key> newItems) {
+    Entity userStats = retreiveUserStats().get();    
+    if (userStats == null) {
```

Comment:
I might be mistaken, but I thought this is guaranteed to not be null if we check the optional return from retrieveUserStats().

`Optional<Entity> userStats = retrieveUserStats()`
`if userStats.isEmpty() { ... }`

Reply:
You're right. I think I removed that part after I committed. I changed it :)

Reply:
You're right, I changed it.